src/Renci.SshNet/ShellStream.cs

```diff
@@ -869,5 +846,37 @@ private void OnDataReceived(byte[] data)
         {
             DataReceived?.Invoke(this, new ShellDataEventArgs(data));
         }
+
+        private string SyncQueuesAndReturn(int bytesToDequeue)
+        {
+            string incomingText;
+
+            lock (_incoming)
+            {
+                var incomingLength = _incoming.Count - _expect.Count + bytesToDequeue;
+                incomingText = _encoding.GetString(_incoming.ToArray(), 0, incomingLength);
+
+                SyncQueuesAndDequeue(bytesToDequeue);
```

Comment:
SyncQueuesAndDequeue method has `lock (_incoming)` and calling is also in the same `lock`. How does it work? I would have thought it would be blocked.

Reply:
Locks can be nested. See [this conversation](https://stackoverflow.com/questions/5035126/why-do-nested-locks-not-cause-a-deadlock).